core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/QueryBindingSetTest.java

```diff
@@ -46,6 +54,136 @@ public void testEqualsMapBindingSet() {
 		assertEquals(qbs, mbs);
 	}
 
+	@Test
+	public void testAddAll() {
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addAll(new QueryBindingSet());
+			assertEquals(0, bs.size());
+		}
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			bs.addAll(new QueryBindingSet());
+			assertEquals(1, bs.size());
+		}
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			final QueryBindingSet bs2 = new QueryBindingSet();
+			bs2.addBinding("foo", vf.createIRI("urn:foo"));
+			bs.addAll(bs2);
+			assertEquals(1, bs.size());
+		}
+		{
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			final QueryBindingSet bs2 = new QueryBindingSet();
+			bs2.addBinding("foo", vf.createIRI("urn:foo"));
+			bs2.addBinding("bar", vf.createIRI("urn:bar"));
+			bs.addAll(bs2);
+			assertEquals(2, bs.size());
+		}
+	}
+
+	@Test
+	public void testAdd() {
+		try {
+			QueryBindingSet bs = new QueryBindingSet();
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			bs.addBinding("foo", vf.createIRI("urn:foo"));
+			fail();
+		} catch (AssertionError e) {
+			return;
+		}
+	}
+
+	@Test
+	public void testNonSharedBackingMap()
+			throws NoSuchFieldException, SecurityException, IllegalArgumentException, IllegalAccessException {
+		QueryBindingSet bs = new QueryBindingSet();
+		bs.addBinding("foo", vf.createIRI("urn:foo"));
+		bs.addBinding("bar", vf.createIRI("urn:bar"));
+		QueryBindingSet bs2 = new QueryBindingSet(bs);
+		final Field declaredField = QueryBindingSet.class.getDeclaredField("bindings");
+		declaredField.setAccessible(true);
+		assertTrue(declaredField.get(bs) != declaredField.get(bs2));
```

Comment:
Rather than testing this through reflection, we could test this through public methods by adding a new binding to bs, after using it to initialize bs2, and then checking that bs2 does not also contain that new binding.